pycodestyle.py

```diff
@@ -296,12 +297,16 @@ def maximum_line_length(physical_line, max_line_length, multiline,
             (len(chunks) == 2 and chunks[0] == '#')) and \
                 len(line) - len(chunks[-1]) < max_line_length - 7:
             return
-        if hasattr(line, 'decode'):   # Python 2
-            # The line could contain multi-byte characters
-            try:
-                length = len(line.decode('utf-8'))
-            except UnicodeError:
-                pass
+        # Special case: multi-byte chars and combining diacritics
+        try:
+            length = sum(not unicodedata.combining(c) for c in line)
+        except TypeError:  # Python 2 str
+            if hasattr(line, 'decode'):  # Python 2 str
+                try:
+                    length = sum(not unicodedata.combining(c)
+                                 for c in line.decode('utf-8'))
+                except UnicodeError:
+                    pass
```

Comment:
I don't think we should be guessing about the types here with `hasattr` and `TypeError` -- let's be explicit and use `sys.version_info >= (3,)` and such

I'd write it like this:

```python
if sys.version_info >= (3,):
    line_text = line
else:
    try:
        line_text = line.decode('UTF-8')
    except UnicodeDecodeError:
        line_text = u''
    length = sum(not unicodedata.combining(c) for c in line)
```

Reply:
I think that a big improvement.